### ISSUE TITLE
Errors in db.transact() closure are not retried

### ISSUE THREAD

**Issue by eastern-grey** [bluejekyll/foundationdb-rs#125]
Hi guys,

I'm seeing an issue where some transactions run with db.transact() are timing out with "Error { kind: \n\nFoundationDB error(1009): Request for future version, should_retry: true }" after 1 second, when a storage server gets a little lagged. I'm puzzled as to why the transaction is not retried, as I'm not setting any transaction options, and I gather there are no timeouts by default. I don't have any experience using foundationdb in other language bindings, so I'm not sure if this is an issue with the rust bindings or in foundationdb itself. If it's the latter, please let me know and I'll take the issue up on the foundationdb forums.

**Comment by eastern-grey** [comment]
It looks like db.transact is only checking the error returned by commit, and not any errors during gets that the closure may return. This seems to be different behaviour to the other language bindings?

The following implementation seems  to work. Not sure if the delays are necessary or whether that's already taken care of by the C bindings. I'm very new to rust, so it probably could be done better.

```rust
pub async fn transact<F, Fut, Output>(db: &Database, func: F) -> Result<Output>
    where
        F: Fn(Transaction) -> Fut,
        Fut: Future<Output = Result<Output>>,
{
    use std::time::Duration;
    use futures_timer::Delay;

    let trx = db.create_trx().unwrap();

    loop {
        let res = func(trx.clone()).await;
        if let Err(e) = res {
            let fail = e.as_fail().find_root_cause();
            if let Some(err) = fail.downcast_ref::<foundationdb::error::Error>() {
                if err.is_retryable() {
                    warn!("retrying error in transaction body: {}", e);
                    Delay::new(Duration::from_millis(100)).await?;
                    continue;
                }
            }

            // return the error
            trx.cancel();
            return Err(e);
        } else {
            // success, commit the transaction
            match trx.clone().commit().compat().await {
                // and return the value from the function on success
                Ok(_) => return res,
                Err(e) => {
                    if e.is_retryable() {
                        warn!("retrying error in transaction commit: {}", e);
                        Delay::new(Duration::from_millis(100)).await?;
                        continue;
                    } else {
                        return Err(e.into());
                    }
                }
            }
        }
    }
}
```

**Comment by pH14** [comment]
Ooh good find, it looks like the commit's result is remapped to the result of the function call. 

https://github.com/bluejekyll/foundationdb-rs/blob/92c30c93eec3714c3015852aa9dc3cf9d8f993df/foundationdb/src/database.rs#L85-L87

I think it's the commit's error that might be swallowed as a result. That probably means a goodly number of errors are not being retried, since so many checks occur at commit time. @eastern-grey interested in tossing up a PR? Since async/await aren't stable yet, we might want to avoid them for now, but the bones of your changes shouldn't be hard to adapt.

**Comment by bluejekyll** [comment]
@pH14 since you've taken such an active role in this (and me less so recently), I've added you as a collaborator. Let me know if you want access to push crates as well.

**Comment by eastern-grey** [comment]
I'm afraid I struggle with the 0.1 future combinators :-(

If my understanding is correct, commit errors are being retried (though without the backoff that on_error() would give), and the issue is that when f() returns an Err, .and_then() never gets called, and the Err is returned to the loop.

Hopefully async is not far off, as it makes things a lot easier. Looks like they're aiming for a beta next month: https://github.com/rust-lang/rust/issues/62149

**Comment by pH14** [comment]
Aha, yes that makes sense, I was misunderstanding how a few of the combinators worked (still a Rust newbie!). An async rewrite of many of these components would be quite nice :)

**Comment by dae** [comment]
I've implemented a tweaked version of this in #135. The on_error() calls are actually unnecessary, as the backoff is applied when a TrxErrFuture is returned.
